core/lib/dal/src/contract_verification_dal.rs

```diff
@@ -560,6 +587,285 @@ impl ContractVerificationDal<'_, '_> {
         .await?
         .flatten())
     }
+
+    async fn get_contract_verification_info_v2(
+        &mut self,
+        address: Address,
+    ) -> anyhow::Result<Option<VerificationInfo>> {
+        // Try perfect match
+        if let Some(info) = sqlx::query!(
+            r#"
+            SELECT
+                verification_info
+            FROM
+                contract_verification_info_v2
+            WHERE
+                initial_contract_addr = $1
+            "#,
+            address.as_bytes(),
+        )
+        .try_map(|row| {
+            serde_json::from_value(row.verification_info).decode_column("verification_info")
+        })
+        .instrument("get_contract_verification_info_v2#perfect_match")
+        .with_arg("address", &address)
+        .fetch_optional(self.storage)
+        .await?
+        .flatten()
+        {
+            return Ok(Some(info));
+        }
+
+        // Try partial match
+        let Some(deployed_contract) = self.get_contract_info_for_verification(address).await?
+        else {
+            return Ok(None);
+        };
+
+        let bytecode_marker = BytecodeMarker::new(deployed_contract.bytecode_hash)
+            .context("unknown bytecode kind")?;
+        let deployed_bytecode = match bytecode_marker {
+            BytecodeMarker::EraVm => deployed_contract.bytecode.as_slice(),
+            BytecodeMarker::Evm => trim_padded_evm_bytecode(
+                BytecodeHash::try_from(deployed_contract.bytecode_hash)
+                    .context("Invalid bytecode hash")?,
+                &deployed_contract.bytecode,
+            )
+            .context("invalid stored EVM bytecode")?,
+        };
+
+        let identifier = ContractIdentifier::from_bytecode(bytecode_marker, deployed_bytecode);
+        // `unwrap_or_default` is safe, since we're checking that `bytecode_without_metadata_keccak256` is not null.
+        let bytecode_without_metadata_keccak256 = identifier
+            .bytecode_without_metadata_keccak256
+            .as_ref()
+            .map(|h| h.hash())
+            .unwrap_or_default();
+
+        let Some((mut info, bytecode_keccak256, bytecode_without_metadata_keccak256)) =
```

Comment:
If similar bytecode is found you can insert info in the DB for this contract. To be honest not sure it will be helpful, it would allow us to easily say how many contracts are verified but we don't really need it. If implemented naively with duplicating verification_info it also will bloat db size, on the other hand splitting the table into two (one with verification_info, bytecode hashes; and another one with verified contract address) will make schema more complex.
So I think the current approach is the right way for now but we can probably consider keeping track of all verified contract addresses explicitly in the future

Reply:
Could you elaborate a bit on how inserting the contract would be helpful?
It won't be accurate, since we will only be doing that if someone queries the contract (so e.g. if nobody checks the contract info, it won't be inserted).
We will save one query at some cases, but at cost of data duplication in the DB.